clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

```diff
@@ -210,6 +220,38 @@ public void testMultipleServerMechanisms() throws Exception {
     }
 
     /**
+     * Tests that Kafka ApiVersionRequests are handled by the SASL server authenticator
+     * prior to SASL handshake flow and that subsequent authentication succeeds.
+     */
+    @Test
+    public void testUnauthenticatedApiVersionRequest() throws Exception {
+        SecurityProtocol securityProtocol = SecurityProtocol.SASL_PLAINTEXT;
+        configureMechanisms("PLAIN", Arrays.asList("PLAIN"));
+        server = NetworkTestUtils.createEchoServer(securityProtocol, saslServerConfigs);
+
+        // Create PLAINTEXT connection to manually authenticate after ApiVersionsRequest
+        String node = "1";
+        createClientConnection(SecurityProtocol.PLAINTEXT, node);
```

Comment:
@ijuma This test uses `PLAINTEXT` for the client connection rather than `SASL_PLAINTEXT` to bypass the authenticators which do the SASL handshake. Instead the test sends the version request and then does the authentication within the test. So even though the client protocol is set to `PLAINTEXT` it simulates `SASL_PLAINTEXT` and talks to a SASL server. I can add a similar one with SSL too.
That reminds me  - I had a question about whether we need the Java client to request ApiVersions before SASL handshake. There is no real reason to do that now since we have only one version of handshake request, but just want to make sure.